pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/PulsarMetadata.java

```diff
@@ -0,0 +1,383 @@
+        if (StringUtils.isBlank(schemaJson)) {
+            throw new PrestoException(NOT_SUPPORTED, "Topic " + topicName.toString()
+                    + " does not have a valid schema");
+        }
+        Schema schema;
+        try {
+            schema = PulsarConnectorUtils.parseSchema(schemaJson);
+        } catch (SchemaParseException ex) {
+            throw new PrestoException(NOT_SUPPORTED, "Topic " + topicName.toString()
+                    + " does not have a valid schema");
+        }
+
+        ImmutableList.Builder<ColumnMetadata> builder = ImmutableList.builder();
+
+        List<Schema.Field> fields = schema.getFields();
+        for (int i = 0; i < fields.size(); i++) {
+            Schema.Field field = fields.get(i);
+            builder.addAll(getColumns(field.name(), field.schema(), i));
+        }
+        if (withInternalColumns) {
+            PulsarInternalColumn.getInternalFields().stream().forEach(new Consumer<PulsarInternalColumn>() {
+                @Override
+                public void accept(PulsarInternalColumn pulsarInternalColumn) {
+                    builder.add(pulsarInternalColumn.getColumnMetadata(false));
+                }
+            });
+        }
+
+        return new ConnectorTableMetadata(schemaTableName, builder.build());
+    }
+
+    private List<PulsarColumnMetadata> getColumns(String name, Schema fieldSchema, int index) {
+
+        List<PulsarColumnMetadata> columnMetadataList = new LinkedList<>();
+
+        if (isPrimitiveType(fieldSchema.getType())) {
+            columnMetadataList.add(new PulsarColumnMetadata(name,
+                    convertType(fieldSchema.getType(), fieldSchema.getLogicalType()),
+                    null, null, false, false, index));
+        } else if (fieldSchema.getType() == Schema.Type.UNION) {
+            boolean canBeNull = false;
+            for (Schema type : fieldSchema.getTypes()) {
+                if (isPrimitiveType(type.getType())) {
+                    PulsarColumnMetadata columnMetadata;
+                    if (type.getType() != Schema.Type.NULL) {
+                        if (!canBeNull) {
+                            columnMetadata = new PulsarColumnMetadata(name,
+                                    convertType(type.getType(), type.getLogicalType()),
+                                    null, null, false, false, index);
+                        } else {
+                            columnMetadata = new PulsarColumnMetadata(name,
+                                    convertType(type.getType(), type.getLogicalType()),
+                                    "field can be null", null, false, false, index);
+                        }
+                        columnMetadataList.add(columnMetadata);
+                    } else {
+                        canBeNull = true;
+                    }
+                }
+            }
+        } else if (fieldSchema.getType() == Schema.Type.RECORD) {
```

Comment:
I guess we are not supported nested types for now? Maybe add a TODO here?

Reply:
yup and sure